src/Database/Query.php

```diff
@@ -237,15 +242,40 @@ public function getConnection(): Connection
      * This method can be overridden in query subclasses to decorate behavior
      * around query execution.
      *
-     * @return \Cake\Database\StatementInterface
+     * @return \Cake\Database\Statement
      */
-    public function execute(): StatementInterface
+    public function execute(): Statement
     {
-        $statement = $this->_connection->run($this);
-        $this->_iterator = $this->_decorateStatement($statement);
+        $this->_statement = $this->_connection->run($this);
         $this->_dirty = false;
 
-        return $this->_iterator;
+        return $this->_statement;
+    }
```

Comment:
This should throw an exception if result decorators were added, but I'm currently calling `execute()` from `all()`.

Reply:
Could we raise an error if `_type !== Query::TYPE_SELECT && $this->_resultDecorators`?

Reply:
I call execute() from all(). I'd have to move this to a protected method.